Sources/SwiftProtobuf/BytecodeReader.swift

```diff
@@ -0,0 +1,139 @@
+    package mutating func nextInstruction() -> Instruction {
+        precondition(hasData, "Unexpected end of bytecode stream")
+
+        let opcode = nextUInt64()
+        precondition(opcode != 0, "Opcode 0 is reserved; do not use it in your own instructions")
+        guard let instruction = Instruction(rawValue: opcode) else {
+            fatalError("Unexpected opcode \(opcode) for instruction set \(Instruction.self)")
+        }
+        return instruction
+    }
+
+    /// Reads and returns the next signed 32-bit integer from the bytecode stream.
+    ///
+    /// This is provided as its own primitive operation because 32-bit values are extremely common
+    /// as field numbers (0 to 2^29-1) and enum cases (-2^31 to 2^31-1). In particular for enum
+    /// cases, using this function specifically for those cases avoids making mistakes involving
+    /// sign- vs. zero-extension between differently-sized integers.
+    /// 
+    /// - Precondition: The reader must not be at the end of the bytecode stream.
+    ///
+    /// - Returns: The signed 32-bit integer that was read from the bytecode stream.
+    package mutating func nextInt32() -> Int32 {
+        // `Int32`s are stored by converting them bit-wise to a `UInt32` and then zero-extended to
+        // `UInt64`, since this representation is smaller than sign-extending them to 64 bits.
+        Int32(bitPattern: UInt32(truncatingIfNeeded: nextUInt64()))
+    }
+
+    /// Reads and returns the next unsigned 64-bit integer from the bytecode stream.
+    ///
+    /// - Precondition: The reader must not be at the end of the bytecode stream.
+    ///
+    /// - Returns: The unsigned 64-bit integer that was read from the bytecode stream.
+    package mutating func nextUInt64() -> UInt64 {
+        precondition(hasData, "Unexpected end of bytecode stream")
+
+        // We store our programs as `StaticString`s, but those are still required to be UTF-8
+        // encoded. This means we can't use a standard varint encoding for integers (because we
+        // cannot arbitrarily use the most significant bit), but we can use a slightly modified
+        // version that always keeps the MSB clear and uses the next-to-MSB as the continuation bit.
+        let byte = UInt64(remainingProgram.first!)
+        remainingProgram = remainingProgram.dropFirst()
+        precondition(byte & 0x80 == 0, "Invalid integer leading byte \(byte)")
+
+        if byte & 0x40 == 0 {
+            return byte
+        }
+        var value: UInt64 = byte & 0x3f
+        var shift: UInt64 = 6
+        while true {
+            let byte = remainingProgram.first!
+            remainingProgram = remainingProgram.dropFirst()
+            value |= UInt64(byte & 0x3f) << shift
+            precondition(byte & 0x80 == 0, "Invalid integer leading byte \(byte)")
+            if byte & 0x40 == 0 {
+                return value
+            }
+            shift += 6
```

Comment:
some guard that `shift` doesn't get too large?

Reply:
Done.